[RFC] cmake: Fix search for msgpack when old system lib is installed

jamessan opened the pull request:
This adds the more thorough version check suggested by @fwalch in #4108.

It also ensures that the header and library built against are from the same
install, instead of picking up the header from the bundled install and the
library from the system install.

I still think it would be a better idea to build bundled dependencies only if
there isn't a sufficient system version installed, but that would require a
larger refactoring.

jamessan committed:
cmake: msgpack: Add more thorough version check

In 33bc332, version constraints were added to pkg_search_module(), but
that only affects the set of directories searched by
find_library()/find_path().

Once the header directory is found, parse the version from
version_master.h so it can be checked by the find_package() call in the
root CMakeLists.txt.

jamessan committed:
cmake: msgpack: Check all lib names per directory

By default, find_library() searches all directories for one possible
name and then looks for the next name.  To make sure we're building
against the same headers and libraries, look for all names in a
directory before moving to the next one.

jamessan commented:
Updated based on @fwalch's comment.

justinmk commented on CMakeLists.txt:
```diff
@@ -279,7 +279,7 @@ include(CheckLibraryExists)
 include_directories(SYSTEM ${LIBUV_INCLUDE_DIRS})
 
-find_package(Msgpack REQUIRED)
+find_package(Msgpack 1.0.0 REQUIRED)
```
So if we want to bump the msgpack version, is this the only place that needs to be updated?

jamessan replied:
Good question.  No, it isn't but I think I see how to make it so that is. :) Currently, FindMsgpack.cmake will also need to update the `pkg_search_module` call, but it looks like `find_package` exports some variables we can use there.

fwalch replied:
The version specified in the `pkg_search_module` call shouldn't be needed anymore, as the version check will be performed by the final `find_package_handle_standard_args`.

jamessan replied:
It's still useful to avoid setting the `PC_*` variables to incorrect paths, since those get used as hints.

jamessan committed:
cmake: Use find_package(Msgpack)'s version in pkg_search_module

Avoid duplicating information by using the Msgpack_FIND_VERSION variable
exported by find_package() inside FindMsgpack's pkg_search_module call.

jamessan commented:
Ok, the version specified in `find_package` is carried through to `pkg_search_module` now.

fwalch commented:
LGTM!
